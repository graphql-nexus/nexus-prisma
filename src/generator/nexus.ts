import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import { GeneratedModule } from './types'

/** Transform the given DMMF into TS source code. */
export function generateRuntimeNexusSource(dmmf: DMMF.Document): GeneratedModule {
  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const content = endent`
    import * as Nexus from 'nexus'

    ${dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)}
  `

  return {
    fileName: 'nexus.ts',
    content,
  }
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`

    export const ${model.name} = {
      ${model.fields.filter(isSupportedType).map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      config: {
        type: ${renderType(field)},
      }
    }
  `
}

function renderType(field: DMMF.Field): string {
  const graphqlType = getGraphQLType(field)

  if (field.isList) {
    return endent`
      Nexus.nonNull(Nexus.list(Nexus.nonNull(${quote(graphqlType)})))
    `
  } else if (field.isRequired) {
    return endent`
      Nexus.nonNull(${quote(graphqlType)})
    `
  } else {
    return endent`
      Nexus.nullable(${quote(graphqlType)})
    `
  }
}

type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'

function quote(x: string): string {
  return `'${x}'`
}

type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

function isSupportedType(field: DMMF.Field): boolean {
  return field.kind === 'scalar'
}

function getGraphQLType(field: DMMF.Field): GraphQLScalarType {
  if (field.kind !== 'scalar') {
    throw new Error(`type mapping only supported for scalar-type fields`)
  }

  const typeName = field.type as PrismaScalarType

  switch (typeName) {
    case 'String': {
      if (field.isId) {
        return GraphQLScalarTypes.ID
      }
      return GraphQLScalarTypes.String
    }
    default: {
      return GraphQLScalarTypes.String
    }
  }
}

const GraphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
} as const
