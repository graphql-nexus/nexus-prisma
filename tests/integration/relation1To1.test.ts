import endent from 'endent'
import { gql } from 'graphql-tag'
import { nonNull, objectType, queryType } from 'nexus'
import { integrationTest } from '../__helpers__'

it('can project user-to-profile relationship', async () => {
  const result = await integrationTest({
    name: `relation1To1-1`,
    datasourceSchema: endent`
      model User {
        id         String    @id
        profile    Profile   @relation(fields: [profileId], references: [id])
        profileId  String
      }
      model Profile {
        id      String  @id
        user    User?
      }
    `,
    datasourceSeed(prisma) {
      return prisma.user.create({
        data: {
          id: 'user1',
          profile: {
            create: {
              id: 'profile1',
            },
          },
        },
      })
    },
    apiSchema({ User, Profile }) {
      return [
        queryType({
          definition(t) {
            t.list.field('users', {
              type: 'User',
              resolve(_, __, ctx) {
                return ctx.prisma.user.findMany()
              },
            })
          },
        }),
        objectType({
          name: User.$name,
          definition(t) {
            t.field(User.id.name, User.id)
            t.field(User.profile.name, User.profile)
          },
        }),
        objectType({
          name: Profile.$name,
          definition(t) {
            t.field(Profile.id.name, Profile.id)
          },
        }),
      ]
    },
    apiClientQuery: gql`
      query {
        users {
          id
          profile {
            id
          }
        }
      }
    `,
  })

  expect(result).toMatchSnapshot()
})

// https://github.com/prisma/nexus-prisma/issues/34
it('can project relationship in opposite direction of where @relation is defined, but the field will be nullable', async () => {
  const result = await integrationTest({
    name: `relation1To1-2`,
    datasourceSchema: endent`
      model User {
        id         String    @id
        profile    Profile   @relation(fields: [profileId], references: [id])
        profileId  String
      }
      model Profile {
        id      String  @id
        user    User?
      }
    `,
    datasourceSeed(prisma) {
      return prisma.user.create({
        data: {
          id: 'user1',
          profile: {
            create: {
              id: 'profile1',
            },
          },
        },
      })
    },
    apiSchema({ User, Profile }) {
      return [
        queryType({
          definition(t) {
            t.list.field('users', {
              type: 'User',
              resolve(_, __, ctx) {
                return ctx.prisma.user.findMany()
              },
            })
          },
        }),
        objectType({
          name: User.$name,
          definition(t) {
            t.field(User.id.name, User.id)
            t.field(User.profile.name, User.profile)
          },
        }),
        objectType({
          name: Profile.$name,
          definition(t) {
            t.field(Profile.id.name, Profile.id)
            t.field(Profile.user.name, Profile.user)
          },
        }),
      ]
    },
    apiClientQuery: gql`
      query {
        users {
          id
          profile {
            id
            user {
              id
            }
          }
        }
      }
    `,
  })

  expect(result).toMatchSnapshot()
})

it('Nullable on Without-Relation-Scalar Side limitation can be worked around by wrapping type in an explicit nonNull', async () => {
  const result = await integrationTest({
    name: `relation1To1-3`,
    datasourceSchema: endent`
      model User {
        id         String    @id
        profile    Profile   @relation(fields: [profileId], references: [id])
        profileId  String
      }
      model Profile {
        id      String  @id
        user    User?
      }
    `,
    datasourceSeed(prisma) {
      return prisma.user.create({
        data: {
          id: 'user1',
          profile: {
            create: {
              id: 'profile1',
            },
          },
        },
      })
    },
    apiSchema({ User, Profile }) {
      return [
        queryType({
          definition(t) {
            t.list.field('users', {
              type: 'User',
              resolve(_, __, ctx) {
                return ctx.prisma.user.findMany()
              },
            })
          },
        }),
        objectType({
          name: User.$name,
          definition(t) {
            t.field(User.id.name, User.id)
            t.field(User.profile.name, User.profile)
          },
        }),
        objectType({
          name: Profile.$name,
          definition(t) {
            t.field(Profile.id.name, Profile.id)
            t.field(Profile.user.name, {
              ...Profile.user,
              type: nonNull(Profile.user.type),
            })
          },
        }),
      ]
    },
    apiClientQuery: gql`
      query {
        users {
          id
          profile {
            id
            user {
              id
            }
          }
        }
      }
    `,
  })

  expect(result).toMatchSnapshot()
})
