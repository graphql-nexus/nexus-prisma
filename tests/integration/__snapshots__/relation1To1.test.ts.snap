// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nullable on Without-Relation-Scalar Side limitation can be worked around by wrapping type in an explicit nonNull 1`] = `
Object {
  "graphqlSchemaSDL": "
type Query {
  users: [User]
}

type User {
  id: ID!
  profile: Profile!
}

type Profile {
  id: ID!
  user: User!
}
",
  "resolution": Object {
    "data": Object {
      "users": Array [
        Object {
          "id": "user1",
          "profile": Object {
            "id": "profile1",
            "user": Object {
              "id": "user1",
            },
          },
        },
      ],
    },
  },
}
`;

exports[`can project relationship in opposite direction of where @relation is defined, but the field will be nullable 1`] = `
Object {
  "graphqlSchemaSDL": "
type Query {
  users: [User]
}

type User {
  id: ID!
  profile: Profile!
}

type Profile {
  id: ID!
  user: User
}
",
  "resolution": Object {
    "data": Object {
      "users": Array [
        Object {
          "id": "user1",
          "profile": Object {
            "id": "profile1",
            "user": Object {
              "id": "user1",
            },
          },
        },
      ],
    },
  },
}
`;

exports[`can project user-to-profile relationship 1`] = `
Object {
  "graphqlSchemaSDL": "
type Query {
  users: [User]
}

type User {
  id: ID!
  profile: Profile!
}

type Profile {
  id: ID!
}
",
  "resolution": Object {
    "data": Object {
      "users": Array [
        Object {
          "id": "user1",
          "profile": Object {
            "id": "profile1",
          },
        },
      ],
    },
  },
}
`;
